lib/compress/index.js

```diff
@@ -5603,10 +5603,12 @@ def_optimize(AST_Call, function(self, compressor) {
             if (has_flag(arg, UNUSED)) continue;
             if (!safe_to_inject
                 || block_scoped.has(arg.name)
-                || identifier_atom.has(arg.name)
-                || scope.conflicting_def(arg.name)) {
+                || identifier_atom.has(arg.name)) {
                 return false;
             }
+            if (scope.conflicting_def(arg.name)) {
+                // TODO arg should be renamed for safety
```

Comment:
I need a function that renames this particular argument in the `fn` function. Is there a method for that?

cc: @fabiosantoscode @alexlamsl

Reply:
It's not just about the one argument in the function.

I've given multiple attempts at this before and I've always encountered a problem like this:

```
function x() {
  const conflict = 4
  function y() {
    z()
  }
  y()
}

const conflict = "PASS"
function z() {
  console.log(conflict)
}
x()
x()
```

In this situation, z() *could* be inlined into y but then it gets the wrong value for `conflict`.

You can reproduce this by running the above through `terser bug.js --toplevel -c defaults=false,inline=true,reduce_vars=true,unused=true`.

For now this is too tricky to fix in a maintainable way, because it involves renaming lots of things and being wary of lots of pitfalls.

Reply:
In my opinion, we should disable inlining altogether if the arguments are not injectable. As I mentioned in #907, this has a worse runtime performance with not much size reduction.


I think we should limit the optimization so only the pure functions get inlined. If a function is pure, there is no need to be worried about its surroundings. We just need to make the inner variables of that pure function unique (for example, by prepending the function name), and inline the code into the call site.

For example, something like this test:
```ts
rename_and_inline: {
    options = {
        toplevel: true,
        defaults: true,
        pure_getters: true
    }
    input: {
        class MyClass {
          method(a) {
           myPureFunction(a)
          }
        }

        function myPureFunction (a) {
         console.log(a)
        }

        new MyClass();
    }
    expect: {
        new class{method(a) {console.log(a)}};
    }
}

```

Reply:
If by "pure" you mean that it doesn't use anything from the closure it should be a safe transform indeed.

Every function has a .clone(deep, toplevel) method, which clones the scope and calls figure_out_scope once more. This cloned function can then get a rename.

There's no function that renames symbols for safety, but there is a method create_symbol which can create a non-colliding symbol taking several scopes into consideration (you want the function scope and the scope you're inlining into).

For renaming you can use walk(node, node => ...) which does a recursive walk, and look for SymbolRef instances that have .definition() === to the definition of the argument you're renaming. You'll have to change the definition's name too, as well as the correct function argument.

Reply:
To address your test, with avoiding inlining when the function is hard to inline, yes, this is very desirable behaviour. Sometimes further passes can make IIFE inlinable, but if this is not frequent (or is fairly predictable) then it's a great improvement.

Reply:
I added a commit. Will this work?
https://github.com/terser/terser/pull/908/commits/247abce36f51d077f87cdb803cf201fae7223f9f

Reply:
I understand the lack of typings is not exactly a plus here :/

I've considered spending time trying to sort that out, but now I've joined an effort to rewrite terser in rust, which will fix that problem once we're done.

Reply:
Writing Terser in a more performant language like Go and Rust would be awesome!